platform-tests/src/test/java/org/junit/platform/launcher/tagexpression/ParserErrorTests.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2015-2017 the original author or authors.
+ *
+ * All rights reserved. This program and the accompanying materials are
+ * made available under the terms of the Eclipse Public License v2.0 which
+ * accompanies this distribution and is available at
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junit.platform.launcher.tagexpression;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+class ParserErrorTests {
+
+	private final Parser parser = new Parser();
+
+	@Test
+	void cantParseExpressionFromNull() {
+		assertThat(parseErrorFromParsing(null)).contains("empty tag expression");
+	}
+
+	@Test
+	void emptyExpression() {
+		assertThat(parseErrorFromParsing("")).contains("empty tag expression");
+	}
+
+	@Test
+	void missingClosingParenthesis() {
+		assertThat(parseErrorFromParsing("(")).contains("( at <0> missing closing parenthesis");
+		assertThat(parseErrorFromParsing("( foo & bar")).contains("( at <0> missing closing parenthesis");
+	}
+
+	@Test
+	void missingOpeningParenthesis() {
+		assertThat(parseErrorFromParsing(")")).contains(") at <0> missing opening parenthesis");
+		assertThat(parseErrorFromParsing(" foo | bar)")).contains(") at <3> missing opening parenthesis");
+	}
+
+	@Test
+	void partialUnaryOperator() {
+		assertThat(parseErrorFromParsing("!")).contains("! at <0> missing rhs operand");
+	}
+
+	@Test
+	void partialBinaryOperator() {
+		assertThat(parseErrorFromParsing("& foo")).contains("& at <0> missing lhs operand");
+		assertThat(parseErrorFromParsing("foo |")).contains("| at <1> missing rhs operand");
```

Comment:
These error messages are already pretty good! I'm wondering whether we could store the character index for each token and emit it as part of the error message instead of the "token index". Thoughts?

Reply:
I'll try to tweak the Lexer and see what I can do.

Reply:
👍

Reply:
Done